### PR TITLE
Allow libraries to reuse binaries compiled with older NIF versions

### PR DESCRIPTION
**Pull request by cocoa-xu**
Hi this PR added a feature that allows libraries to reuse binaries compiled with older NIF versions instead of compiling the same code over and over again for all NIF versions they can support.

The main motivation is to reduce CI build time and release size as older NIF libraries can be used with newer Erlang/OTP versions.

This feature is done by changing how users can config the `:make_precompiler_nif_versions` key. And now there are two ways configure `elixir_make` on how to compile and reuse precompiled binaries. 

##### Use minimum NIF version (recommended)
The first way is to use the minimum NIF version that is available to the current target. For example, if the NIF library
only uses functions and feature available in NIF version `2.14`, the following configuration should be set:

```elixir
make_precompiler_nif_versions: [
  use_minimum_version: true,
  minimum_version: "2.14"
]
```

In this way, we can reuse the same precompiled binary for newer Erlang/OTP versions instead of compiling the same code over and over again. And if the user is using the NIF library on an older Erlang/OTP version, `elixir_make` will try to compile the NIF library for the target.

Additionally, if the NIF library can optionally use functions and/or features that are only available in NIF version `2.17`, then the following configuration can be set to provide better support for newer Erlang/OTP versions while still providing support for older Erlang/OTP versions:

```elixir
make_precompiler_nif_versions: [
  use_minimum_version: true,
  minimum_version: fn _target, current_nif_version ->
    if current_nif_version >= "2.17" do
        "2.17"
    else
        "2.14"
    end
  end
]
```

The above example tells `elixir_make` that it can reuse binaries compiled with NIF version `2.17` if the current host supports it; otherwise, it will use binaries compiled with NIF version `2.14`.

Users can further change the above example to use different target name to control which feature set is available. For example, to use NIF version `2.17` for any `aarch64` targets and `2.14` for other targets:

```elixir
make_precompiler_nif_versions: [
  use_minimum_version: true,
  minimum_version: fn target, current_nif_version ->
    if current_nif_version >= "2.17" and String.contains?(target, "aarch64") do
        "2.17"
    else
        "2.14"
    end
  end
]
```

However, the `versions_for_target` sub-key must be set to inform `:elixir_make` that the precompiled artefacts may have different NIF versions for different targets. For example,

```elixir
make_precompiler_nif_versions: [
  use_minimum_version: true,
  minimum_version: fn target, current_nif_version ->
    if current_nif_version >= "2.17" and String.contains?(target, "aarch64") do
        "2.17"
    else
        "2.14"
    end
  end,
  versions_for_target: fn target ->
    if String.contains?(target, "aarch64") do
      ["2.17", "2.14"]
    else
      ["2.14"]
    end
  end
]
```

This information is used to determine all available precompiled artefacts and is used to fetch all of them and generate the checksum file.

```elixir
make_precompiler_nif_versions: [
  use_minimum_version: true,
  minimum_version: fn target, current_nif_version ->
    if current_nif_version >= "2.17" do
        "2.17"
    else
        "2.14"
    end
  end
]
```

##### Use exact NIF version
The second (and the old) way is to use the exact NIF version that is available to the current target. In this case, a list of supported NIF versions should be set in key `versions`. For example, 

```elixir
make_precompiler_nif_versions: [versions: ["2.14", "2.15", "2.16"]]
```

The above example tells `elixir_make` that precompiled artefacts are available for these NIF versions. If there's no precompiled artefacts available on the current host, `elixir_make` will try to compile the NIF library.

If you'd like to aim for an older NIF version, say `2.15` for Erlang/OTP 23 and 24, then you need to setup CI correspondingly and set the value of this key to `[versions: ["2.15", "2.16"]]`. This optional key will only be checked when downloading precompiled artefacts.

For some platforms maybe we only have precompiled artefacts after a certain NIF version, say for x86_64 Windows we have precompiled artefacts available when NIF version >= `2.16` while other platforms have precompiled artefacts available from NIF version >= `2.15`.

In such case we can inform `:elixir_make` that Windows targets don't have precompiled artefacts available except for NIF version `2.16` by passing a function to the `availability` sub-key.

```elixir
defp target_available_for_nif_version?(target, nif_version) do
  if String.contains?(target, "windows") do
    nif_version == "2.16"
  else
    true
  end
end
```

The default value for `make_precompiler_nif_versions` is 

```elixir
[versions: ["#{:erlang.system_info(:nif_version)}"]]
```

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you @cocoa-xu! I think this is an important feature to have indeed.

I was thinking though, perhaps we should have it enabled by default. This means that, as soon as we release a new version of `elixir_make`, everyone can benefit from it without publishing new assets. The default behaviour would be to find biggest major version still supported. For example, this:

```
make_precompiler_nif_versions: [
  versions: ["2.14", "2.15"],
]
```

Now imagine that "2.16" is released, we can do this:

```
defp nif_version_to_tuple(nif_version) do
  [major, minor | _] = String.split(nif_version, ".")
  {String.to_integer(major), String.to_integer(minor)}
end

defp find_nif_version(nif_version, versions) do
  if nif_version in versions do
    # use it
  else
    # default fallback logic
    {major, minor} = nif_version_to_tuple(nif_version)
    versions = Enum.map(versions, &nif_version_to_tuple/1)

    # Get all matching major versions, earlier than the current version
    # and their distance. We want the closest (smallest distance).
    candidates =
      for version <- versions,
          {^major, candidate_minor} <- [nif_version_to_tuple(version)],
          candidate_minor <= minor,
          do: {minor - candidate_minor, version}

    case Enum.sort(candidates) do
      [{_, version} | _] -> version
      _ -> # no candidates
    end
  end
end
```

The second part of the patch is to make it customizable. I think we can provide a `:fallback_version` option, which is pretty much the same option as you defined for `minimum_version`, except with a different name and it is only invoked in case we can't find a matching name:

```
make_precompiler_nif_versions: [
  fallback_version: fn target, current_nif_version ->
    if current_nif_version >= "2.17" and String.contains?(target, "aarch64") do
        "2.17"
    else
        "2.14"
    end
  end
]
```

WDYT? Would this work?

**Comment by josevalim** [comment]
Well, my snippets above need to be slightly modified to only pass versions that are available. Perhaps, the `fallback_version` function should receive three arguments `target, current_nif_version, available_nif_versions` (which are all versions after the availability check).

**Comment by cocoa-xu** [comment]
Hi @josevalim, I think the suggested changes should work. I've pushed some new commits to do it. Probably we can deprecate the availability key in this PR, and change how we use the `versions` key in #80?

File: lib/elixir_make/artefact.ex [hunk 1]
```diff
@@ -66,12 +66,26 @@ defmodule ElixirMake.Artefact do
   end
 
   defp archive_filename(config, target, nif_version) do
+    precompiler_archive_filename = config[:make_precompiler_archive_filename] || :default
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:69]: Should we revert this change? It is not documented or used anywhere it seems.

File: lib/elixir_make/artefact.ex [hunk 2]
```diff
@@ -151,33 +190,32 @@ defmodule ElixirMake.Artefact do
...
 
+    current_nif_version = "#{:erlang.system_info(:nif_version)}"
+
     nif_versions =
       config[:make_precompiler_nif_versions] ||
-        [versions: ["#{:erlang.system_info(:nif_version)}"]]
+        [versions: [current_nif_version]]
+
+    versions = nif_versions[:versions]
+    versions_for_target = nif_versions[:versions_for_target]
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/elixir_make/artefact.ex:200]: And we cannot remove availability yet, it is used by sqlite. So it would be a breaking change. :(

**Comment by josevalim** [comment]
Thank you @cocoa-xu!

> Probably we can deprecate the availability key in this PR, and change how we use the versions key in https://github.com/elixir-lang/elixir_make/issues/80?

I would prefer to keep the pull request smaller. So I would add fallback version to this PR and we deprecate availability and add further features later. This way we can discuss each change individually, instead of juggling several at once. :)

**Comment by cocoa-xu** [comment]
> Thank you @cocoa-xu!
> 
> > Probably we can deprecate the availability key in this PR, and change how we use the versions key in #80?
> 
> I would prefer to keep the pull request smaller. So I would add fallback version to this PR and we deprecate availability and add further features later. This way we can discuss each change individually, instead of juggling several at once. :)

No problem, I've reverted related changes and only added `fallback_version` for this PR/

File: lib/elixir_make/artefact.ex [hunk 3]
```diff
@@ -187,22 +216,48 @@ defmodule ElixirMake.Artefact do
...
+            :no_candidates ->
+              if is_function(fallback_version, 3) do
+                fallback_version.(current_target, current_nif_version, versions)
+              else
+                current_nif_version
+              end
+
+            version when is_binary(version) ->
+              version
+          end
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:240]: I was thinking more something like this:

```suggestion
        fallback_version = nif_versions[:fallback_version] || &find_nif_version/3
        nif_version_to_use = fallback_version.(current_target, current_nif_version, versions)
```

This way someone has full control over the algorithm. But maybe that's too flexible? Do you think someone would even want to customize the default behaviour?

File: lib/elixir_make/artefact.ex [hunk 4]
```diff
@@ -187,22 +216,48 @@ defmodule ElixirMake.Artefact do
...
 
-        case List.keyfind(available_urls, target_at_nif_version, 0) do
-          {^target_at_nif_version, download_url} ->
-            {:ok, current_target, download_url}
+        case nif_version_to_use do
+          :compile ->
+            {:error, :compile_from_source}
+
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:245]: Do you think we need this clause? If we don't have this clause, it would return `{:error, {:unavailable_target, ...}}`, which would translate to a compile anyway.

Comment (cocoa-xu) [reply lib/elixir_make/artefact.ex:245]: I was thinking about this too. Adding this was to suppress the warning message below, but I guess it's not really a problem if we show this in the log?

Comment (josevalim) [reply lib/elixir_make/artefact.ex:233]: I think the error message is correct, that we indeed did not find something precompiled, and that you need to compile from source. I think letting the user know there is no precompiled is useful feedback? If someone complains add the feature later :D

File: lib/elixir_make/artefact.ex (hunk 3 above)
Comment (cocoa-xu) [reply lib/elixir_make/artefact.ex]: Ah, maybe I was overthinking regarding how users may want to use this feature. It should be already flexible enough for most cases. I'll apply the suggestion.

File: lib/elixir_make/artefact.ex [hunk 5]
```diff
@@ -187,22 +216,36 @@ defmodule ElixirMake.Artefact do
...
       {:ok, current_target} ->
-        available_urls = available_target_urls(config, precompiler)
-        target_at_nif_version = {current_target, nif_version}
+        nif_versions =
+          config[:make_precompiler_nif_versions] ||
+            [versions: []]
+
+        versions = nif_versions[:versions]
+        fallback_version = nif_versions[:fallback_version] || &find_nif_version/3
+        nif_version_to_use = fallback_version.(current_target, current_nif_version, versions)
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:228]: Sorry, we wqill need to change `find_nif_version/3` accordingly, to receive an extra argument and to return the `current_nif_version` instead of `:no_candiates`. :D

File: lib/elixir_make/artefact.ex (hunk 4 above)
Comment (cocoa-xu) [reply lib/elixir_make/artefact.ex]: No problem~ I'll change this back!

File: lib/elixir_make/artefact.ex (hunk 5 above)
Comment (cocoa-xu) [reply lib/elixir_make/artefact.ex:228]: Okay, it's done. ;)

File: PRECOMPILATION_GUIDE.md [hunk 6]
```diff
@@ -105,17 +105,71 @@ Directory structures and symbolic links are preserved.
...
+The third optional config key is `make_precompiler_nif_versions`, which configures `elixir_make` on how to compile and reuse precompiled binaries.
+
+The default value for `make_precompiler_nif_versions` is 
 
 ```elixir
 [versions: ["#{:erlang.system_info(:nif_version)}"]]
 ```
 
-If you'd like to aim for an older NIF version, say `2.15` for Erlang/OTP 23 and 24, then you need to setup CI correspondingly and set the value of this key to `[versions: ["2.15", "2.16"]]`. This optional key will only be checked when downloading precompiled artefacts.
+There're two sub-keys for `make_precompiler_nif_versions`:
```
Comment (josevalim) [inline PRECOMPILATION_GUIDE.md:116]:
```suggestion
There're three sub-keys for `make_precompiler_nif_versions`:
```

File: lib/elixir_make/artefact.ex [hunk 7]
```diff
@@ -141,6 +141,31 @@ defmodule ElixirMake.Artefact do
...
+  defp nif_version_to_tuple(nif_version) do
+    [major, minor | _] = String.split(nif_version, ".")
+    {String.to_integer(major), String.to_integer(minor)}
+  end
+
+  defp find_nif_version(_current_target, current_nif_version, versions) do
+    if current_nif_version in versions do
+      current_nif_version
+    else
+      {major, minor} = nif_version_to_tuple(current_nif_version)
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:153]:
```suggestion
      {major, minor} = nif_version_to_tuple(current_nif_version)

```

File: lib/elixir_make/artefact.ex [hunk 8]
```diff
@@ -187,11 +217,19 @@ defmodule ElixirMake.Artefact do
...
+  def current_target_url(config, precompiler, current_nif_version) do
     case precompiler.current_target() do
       {:ok, current_target} ->
+        nif_versions =
+          config[:make_precompiler_nif_versions] ||
+            [versions: []]
+
+        versions = nif_versions[:versions]
+        fallback_version = nif_versions[:fallback_version] || (&find_nif_version/3)
+        nif_version_to_use = fallback_version.(current_target, current_nif_version, versions)
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:229]:
```suggestion

        nif_version_to_use =
          if current_nif_version in versions do
            current_nif_version
          else
            fallback_version = nif_versions[:fallback_version] || (&fallback_version/3)
            fallback_version.(current_target, current_nif_version, versions)
          end
```

File: lib/elixir_make/artefact.ex [hunk 9]
```diff
@@ -141,6 +141,32 @@ defmodule ElixirMake.Artefact do
...
+            {^major, candidate_minor} <- [nif_version_to_tuple(version)],
+            candidate_minor <= minor,
+            do: {minor - candidate_minor, version}
+
+      case Enum.sort(candidates) do
+        [{_, version} | _] -> version
+        _ -> current_nif_version
+      end
+    end
+  end
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:168]:
```suggestion
  defp fallback_version(_current_target, current_nif_version, versions) do
    {major, minor} = nif_version_to_tuple(current_nif_version)

    # Get all matching major versions, earlier than the current version
    # and their distance. We want the closest (smallest distance).
    candidates =
      for version <- versions,
          {^major, candidate_minor} <- [nif_version_to_tuple(version)],
          candidate_minor <= minor,
          do: {minor - candidate_minor, version}

    case Enum.sort(candidates) do
      [{_, version} | _] -> version
      _ -> current_nif_version
    end
  end
```

File: lib/elixir_make/artefact.ex [hunk 10]
```diff
@@ -187,11 +213,25 @@ defmodule ElixirMake.Artefact do
...
+
+        versions = nif_versions[:versions]
+
+        nif_version_to_use =
+          if current_nif_version in versions do
+            current_nif_version
+          else
+            fallback_version = nif_versions[:fallback_version] || (&fallback_version/3)
+            fallback_version.(current_target, current_nif_version, versions)
+          end
```
Comment (josevalim) [inline lib/elixir_make/artefact.ex:231]: @cocoa-xu I just moved the conditional outside. I think that, if the `current_nif_version` matches, we should never invoke the fallback. Makes sense right? If you are happy, feel free to merge it, otherwise I do it when I am back home!

Comment (josevalim) [reply lib/elixir_make/artefact.ex:231]: Oh, I thought this was a elixir-nx project, it isn't. :) I will go ahead and merge it.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/elixir_make/artefact.ex (hunk 10 above)
Comment (cocoa-xu) [reply lib/elixir_make/artefact.ex:231]: > @cocoa-xu I just moved the conditional outside. I think that, if the `current_nif_version` matches, we should never invoke the fallback. Makes sense right? If you are happy, feel free to merge it, otherwise I do it when I am back home!

Ah you're right!

> Oh, I thought this was a elixir-nx project, it isn't. :) I will go ahead and merge it.

:D
